numba/runtime/_nrt_python.c

```diff
@@ -0,0 +1,539 @@
+#include "_pymodule.h"
+#include "nrt.h"
+
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/ndarrayobject.h>
+#include <numpy/arrayscalars.h>
+
+#include "_arraystruct.h"
+
+/* For Numpy 1.6 */
+#ifndef NPY_ARRAY_BEHAVED
+    #define NPY_ARRAY_BEHAVED NPY_BEHAVED
+#endif
+
+
+static
+PyObject*
+memsys_shutdown(PyObject *self, PyObject *args) {
+    if (!PyArg_ParseTuple(args, "")) {
+        return NULL;
+    }
+    NRT_MemSys_shutdown();
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+memsys_set_atomic_inc_dec(PyObject *self, PyObject *args) {
+    unsigned PY_LONG_LONG addr_inc, addr_dec;
+    if (!PyArg_ParseTuple(args, "KK", &addr_inc, &addr_dec)) {
+        return NULL;
+    }
+    NRT_MemSys_set_atomic_inc_dec((void*)addr_inc, (void*)addr_dec);
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+memsys_set_atomic_cas(PyObject *self, PyObject *args) {
+    unsigned PY_LONG_LONG addr_cas;
+    if (!PyArg_ParseTuple(args, "K", &addr_cas)) {
+        return NULL;
+    }
+    NRT_MemSys_set_atomic_cas((void*)addr_cas);
+    Py_RETURN_NONE;
+}
+
+static
+PyObject*
+memsys_process_defer_dtor(PyObject *self, PyObject *args) {
+    if (!PyArg_ParseTuple(args, "")) {
+        return NULL;
+    }
+    NRT_MemSys_process_defer_dtor();
+    Py_RETURN_NONE;
+}
+
+
+static
+void pyobject_dtor(void *ptr, void* info) {
+    PyGILState_STATE gstate;
+    PyObject *ownerobj = info;
+
+    gstate = PyGILState_Ensure();   /* ensure the GIL */
```

Comment:
If the destructors are always "deferred", you shouldn't need this (i.e. the destructor is called either from the wrapper or Python itself).